### PR TITLE
Clean up of method reference and lambda code

### LINKED ISSUE #1319: Method reference type inference warning given despite lack of inference
This code produces a `methodref.inference.unimplemented` warning (as in #802):

```java
java.util.stream.Stream<String> stream = java.util.stream.Stream.of("foo");
Iterable<String> iterable = stream::iterator;
```

```
warning: [methodref.inference.unimplemented] This version of the Checker Framework does not type-check method references with implicit type arguments.
Iterable<String> iterable = stream::iterator;
                            ^
```

(It does not matter what the stream’s actual element type is or how the stream is obtained.)

I don’t understand how this warning is relevant to this line or why a warning is given at all.

I’ve tried Checker Framework 2.1.12 and 2.1.11, both with Oracle’s javac 1.8.0_40.

### PR DESCRIPTION
**Pull request by smillst**
1. Use variable names that correspond better to terms in the JLS.
2. Add documentation
3. Add "inference" for one particular kind of method reference (See checker/tests/nullness/java8/methodref/ClassTypeArgInference.java for examples.)
4. Correct and explain type of lambda parameters (See checker/tests/nullness/java8/lambda/ParametersInBodyGenerics.java for examples.)
5. Correct code that decides if a method reference requires type argument inference. (Fixes #1319.)

### CODE REVIEW

File: checker/tests/nullness/java8/methodref/ClassTypeArgInference.java [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+        Gen<String> o = new Gen<>("");
+        // :: error: (methodref.param.invalid)
+        Factory f = Gen::make;
+        // :: error: (methodref.param.invalid)
+        Factory f2 = Gen<String>::make;
+        // :: error: (methodref.receiver.invalid) :: error: (methodref.return.invalid)
+        Factory f3 = Gen<@Nullable String>::make;
+        f2.make(o, null).toString();
+
+        /*  Factory2 factory2a = Gen<@Nullable String>::new;
```
Comment (wmdietl) [inline checker/tests/nullness/java8/methodref/ClassTypeArgInference.java]: What do these comments mean? Are these more tests that should be enabled?
Class Factory2 is currently unused.

File: framework/src/org/checkerframework/common/basetype/BaseTypeVisitor.java [hunk 2]
```diff
@@ -2489,120 +2492,135 @@ protected boolean checkMethodReferenceAsOverride(
 
         Pair<AnnotatedDeclaredType, AnnotatedExecutableType> result =
                 atypeFactory.getFnInterfaceFromTree(memberReferenceTree);
-        AnnotatedDeclaredType overriddenType = result.first;
-        AnnotatedExecutableType overriddenMethodType = result.second;
+        // The type to which the member reference is a assigned -- also known as the target type of the reference.
```
Comment (wmdietl) [inline framework/src/org/checkerframework/common/basetype/BaseTypeVisitor.java]: `a assigned` -> `assigned`

File: framework/src/org/checkerframework/common/basetype/BaseTypeVisitor.java [hunk 3]
```diff
@@ -2489,120 +2492,135 @@ protected boolean checkMethodReferenceAsOverride(
...
+        AnnotatedDeclaredType functionalInterface = result.first;
+        // The type of the single method in declared by the functional interface.
```
Comment (wmdietl) [inline framework/src/org/checkerframework/common/basetype/BaseTypeVisitor.java]: The `in` doesn't sound right. Should it be 'that is`?

File: framework/src/org/checkerframework/common/basetype/BaseTypeVisitor.java [hunk 4]
```diff
@@ -2489,120 +2492,135 @@ protected boolean checkMethodReferenceAsOverride(
...
 
         // ========= Overriding Type =========
-        // Get declared type from <expression>::method or <type use>::method
         // This doesn't get the correct type for a "MyOuter.super" based on the receiver of the
         // enclosing method.
         // That is handled separately in method receiver check.
-        // TODO: Class type argument inference
-        AnnotatedTypeMirror overridingType =
+
+        // The type of from  <expression>::method or <type use>::method.
```
Comment (wmdietl) [inline framework/src/org/checkerframework/common/basetype/BaseTypeVisitor.java]: `of from` sounds strange.

File: framework/src/org/checkerframework/common/basetype/BaseTypeVisitor.java [hunk 5]
```diff
@@ -2489,120 +2492,135 @@ protected boolean checkMethodReferenceAsOverride(
...
+        AnnotatedTypeMirror type =
```
Comment (wmdietl) [inline framework/src/org/checkerframework/common/basetype/BaseTypeVisitor.java]: `type` is a very generic name. Can you find something more descriptive?

File: framework/src/org/checkerframework/common/basetype/BaseTypeVisitor.java [hunk 6]
```diff
@@ -2489,120 +2492,135 @@ protected boolean checkMethodReferenceAsOverride(
...
                 atypeFactory.getAnnotatedType(memberReferenceTree.getQualifierExpression());
 
         // ========= Overriding Executable =========
         // The ::method element
-        ExecutableElement overridingElement =
+        ExecutableElement compileTimeDeclaration =
```
Comment (wmdietl) [inline framework/src/org/checkerframework/common/basetype/BaseTypeVisitor.java]: How about `memberReferenceElement`?
All of this happens at compile time...

File: framework/src/org/checkerframework/common/basetype/BaseTypeVisitor.java [hunk 7]
```diff
@@ -2489,120 +2492,135 @@ protected boolean checkMethodReferenceAsOverride(
...
                 (ExecutableElement) TreeUtils.elementFromTree(memberReferenceTree);
-        AnnotatedExecutableType overridingMethodType =
-                atypeFactory.methodFromUse(memberReferenceTree, overridingElement, overridingType)
-                        .first;
 
-        if (checkMethodReferenceInference(
-                memberReferenceTree, overridingMethodType, overriddenMethodType, overridingType)) {
+        if (type.getKind() == TypeKind.DECLARED && ((AnnotatedDeclaredType) type).wasRaw()) {
+            if (((JCMemberReference) memberReferenceTree).kind == ReferenceKind.UNBOUND) {
+                // The method reference is of the form :Type # instMethod
```
Comment (wmdietl) [inline framework/src/org/checkerframework/common/basetype/BaseTypeVisitor.java]: Spacing between `form: Type`?

File: framework/src/org/checkerframework/common/basetype/BaseTypeVisitor.java [hunk 8]
```diff
@@ -2489,120 +2492,135 @@ protected boolean checkMethodReferenceAsOverride(
...
+                if (asSuper != null) {
+                    type = AnnotatedTypes.asSuper(atypeFactory, p1, type);
+                }
+            }
+            // else method reference is something like ArrayList::new
+            // TODO: Use diamond, <>, inference to infer the class type arguments.
+            // for now this case is skipped below in checkMethodReferenceInference.
+        }
+
+        // They type of the compileTimeDeclaration if it were invoked with a receiver expression
```
Comment (wmdietl) [inline framework/src/org/checkerframework/common/basetype/BaseTypeVisitor.java]: `They` -> `The`

File: framework/src/org/checkerframework/common/basetype/BaseTypeVisitor.java [hunk 9]
```diff
@@ -2489,120 +2492,135 @@ protected boolean checkMethodReferenceAsOverride(
...
-            overriddenMethodType = overriddenMethodType.getErased();
+        // This needs to be done before invocationType.getReturnType() and
+        // functionType.getReturnType()
+        if (invocationType.getTypeVariables().isEmpty()
+                && !functionType.getTypeVariables().isEmpty()) {
+            functionType = functionType.getErased();
         }
 
-        // Use the functional interface's parameters to resolve poly quals.
+        // Use the function type's parameters to resolve polymorpich qualifiers.
```
Comment (wmdietl) [inline framework/src/org/checkerframework/common/basetype/BaseTypeVisitor.java]: `polymorpich` -> `polymorphic`

File: framework/src/org/checkerframework/common/basetype/BaseTypeVisitor.java [hunk 10]
```diff
@@ -2489,120 +2492,135 @@ protected boolean checkMethodReferenceAsOverride(
...
-
-                requiresInference = true;
-            } else if (overridingType.getKind() == TypeKind.DECLARED
-                    && ((AnnotatedDeclaredType) overridingType).getTypeArguments().size() > 0) {
+        boolean requiresInference = false;
+        // If the function to which the member reference refers is generic, but the member
+        // reference does not provide method type arguments, then java 8 inference is required.
+        // Issue 979.
+        if (invocationType.getTypeVariables().size() > 0
+                && (memberReferenceTree.getTypeArguments() == null
```
Comment (wmdietl) [inline framework/src/org/checkerframework/common/basetype/BaseTypeVisitor.java]: Can this ever return null?

File: framework/src/org/checkerframework/common/basetype/BaseTypeVisitor.java [hunk 11]
```diff
@@ -2489,120 +2492,135 @@ protected boolean checkMethodReferenceAsOverride(
...
+                        || memberReferenceTree.getTypeArguments().size() == 0)) {
```
Comment (wmdietl) [inline framework/src/org/checkerframework/common/basetype/BaseTypeVisitor.java]: `.size() == 0` -> `.isEmpty()`

File: framework/src/org/checkerframework/framework/type/TypeFromMemberVisitor.java [hunk 12]
```diff
@@ -86,10 +96,30 @@ private static void inferLambdaParamAnnotations(
...
-            AnnotatedExecutableType fnMethod = res.second;
-            AnnotatedTypeMirror declaredParam = fnMethod.getParameterTypes().get(index);
-            // TODO: Should we infer nested types (e.g. List<@x String>)
-            result.addMissingAnnotations(declaredParam.getAnnotations());
+            AnnotatedExecutableType functionType = res.second;
+            AnnotatedTypeMirror funcTypeParam = functionType.getParameterTypes().get(index);
+            if (TreeUtils.isImplicitlyTypeLambda(declaredInTree)) {
+                if (TypesUtils.isErasedSubtype(
+                        funcTypeParam.actualType, lambdaParam.actualType, f.types)) {
+                    // The Java types should be exactly the same, but because invocation type inferece
```
Comment (wmdietl) [inline framework/src/org/checkerframework/framework/type/TypeFromMemberVisitor.java]: `inferece` -> `inference`

File: framework/src/org/checkerframework/framework/type/TypeFromMemberVisitor.java [hunk 13]
```diff
@@ -86,10 +96,30 @@ private static void inferLambdaParamAnnotations(
...
+                    // (#979) isn't implement, check first.
+                    return AnnotatedTypes.asSuper(f, funcTypeParam, lambdaParam);
+                }
+                lambdaParam.addMissingAnnotations(funcTypeParam.getAnnotations());
+                return lambdaParam;
+
+            } else {
+                // The lambda expressions is explicitly type, so the parameters have declared types:
```
Comment (wmdietl) [inline framework/src/org/checkerframework/framework/type/TypeFromMemberVisitor.java]: `expressions` -> `expression`
`type` -> `typed`

File: framework/src/org/checkerframework/framework/type/TypeFromMemberVisitor.java [hunk 14]
```diff
@@ -86,10 +96,30 @@ private static void inferLambdaParamAnnotations(
...
+                // (String s) -> ...
+                // The declared type that may or may not have explicit annotations.
```
Comment (wmdietl) [inline framework/src/org/checkerframework/framework/type/TypeFromMemberVisitor.java]: Should the `that` be removed?

File: javacutil/src/org/checkerframework/javacutil/TreeUtils.java [hunk 15]
```diff
@@ -1279,4 +1281,18 @@ public static Symbol findFunction(Tree tree, ProcessingEnvironment env) {
         Types javacTypes = Types.instance(ctx);
         return javacTypes.findDescriptorSymbol(((Type) typeOf(tree)).asElement());
     }
+
+    /**
+     * Returns true if {@code tree} is an implicitly typed lambda.
+     *
+     * <p>An lambda whose formal type parameters have inferred types is an implicitly typed lambda.
```
Comment (wmdietl) [inline javacutil/src/org/checkerframework/javacutil/TreeUtils.java]: `An lambda` -> `A lambda expression`

File: framework/src/org/checkerframework/common/basetype/BaseTypeVisitor.java (hunk 10 above)
Comment (smillst) [reply framework/src/org/checkerframework/common/basetype/BaseTypeVisitor.java]: Yes, it's null when type argument inference is required.

File: javacutil/src/org/checkerframework/javacutil/TreeUtils.java [hunk 16]
```diff
@@ -1279,4 +1281,18 @@ public static Symbol findFunction(Tree tree, ProcessingEnvironment env) {
...
+    /**
+     * Returns true if {@code tree} is an implicitly typed lambda.
+     *
+     * <p>A lambda expression whose formal type parameters have inferred types is an implicitly
+     * typed lambda. (See JLS 15.27.1)
+     *
+     * @param tree any kind of tree
+     * @return true iff {@code tree} is an implicitly typed lambda.
+     */
+    public static boolean isImplicitlyTypeLambda(Tree tree) {
```
Comment (wmdietl) [inline javacutil/src/org/checkerframework/javacutil/TreeUtils.java]: `isImplicitlyTypeLambda` -> `isImplicitlyTypedLambda` for consistency.

Comment (smillst) [reply javacutil/src/org/checkerframework/javacutil/TreeUtils.java]: Thanks for catching all my typos!

**Pull request merged**